_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw_utils.cpp

```diff
@@ -1122,22 +1122,80 @@ namespace MfxHwH264EncodeHW
 using namespace MfxHwH264EncodeHW;
 inline void SetMinMaxQP(mfxExtCodingOption2 const &  extOpt2, mfxU8  QPMin[], mfxU8  QPMax[])
 {
-	// qp=0 doesn't supported
-	QPMin[0] = (extOpt2.MinQPI) ? extOpt2.MinQPI : 1;
-	QPMin[1] = (extOpt2.MinQPP) ? extOpt2.MinQPP : 1;
-	QPMin[2] = (extOpt2.MinQPB) ? extOpt2.MinQPB : 1;
+	// valid qp range [1, 51], default qp range [8, 51], qp=0 doesn't supported by driver
+	QPMin[0] = (extOpt2.MinQPI) ? extOpt2.MinQPI : 8;
+	QPMin[1] = (extOpt2.MinQPP) ? extOpt2.MinQPP : 8;
+	QPMin[2] = (extOpt2.MinQPB) ? extOpt2.MinQPB : 8;
 
 	QPMax[0] = (extOpt2.MaxQPI) ? extOpt2.MaxQPI : 51;
 	QPMax[1] = (extOpt2.MaxQPP) ? extOpt2.MaxQPP : 51;
 	QPMax[2] = (extOpt2.MaxQPB) ? extOpt2.MaxQPB : 51;
 }
 
+enum
+{
+    MFX_BRC_ACCURACY_UNKNOWN,
+    MFX_BRC_ACCURACY_LOW,
+    MFX_BRC_ACCURACY_MEDIUM,
+    MFX_BRC_ACCURACY_HIGH
+};
+
+
+struct sBrcTarget
+{
+    mfxU32 rateCalcPeriodInSec; //  0 - all frames are used in rate calculation, otherwise the last n seconds
+    mfxU32 accuracy;            //  MFX_BRC_ACCURACY_LOW, MFX_BRC_ACCURACY_MEDIUM, MFX_BRC_ACCURACY_HIGH
+};
+
+
+
+void setLAThresholds(sLAThresholds& thresholds, mfxU32 accuracy)
+{
+    thresholds.minFramesForClassicLA = 30;
+    thresholds.minFramesForStat = 10;
+
+    if (accuracy == MFX_BRC_ACCURACY_HIGH)
+    {
+        thresholds.minCostCalcPeriod = 20;
+        thresholds.maxRateRatioLocal = 1.3;
+        thresholds.minRateRatioLocal = 0.9;
+        thresholds.maxAvgRateRatio = 1.02;
+        thresholds.minAvgRateRatio = 1.00;
+    }
+    else if (accuracy == MFX_BRC_ACCURACY_LOW)
+    {
+        thresholds.minCostCalcPeriod = 40;
+        thresholds.maxRateRatioLocal = 1.5;
+        thresholds.minRateRatioLocal = 0.7;
+        thresholds.maxAvgRateRatio = 1.08;
+        thresholds.minAvgRateRatio = 1.00;
+    }
+    else
+    {
+        // MFX_BRC_ACCURACY_MEDIUM
+        thresholds.minCostCalcPeriod = 30;
+        thresholds.maxRateRatioLocal = 1.4;
+        thresholds.minRateRatioLocal = 0.85;
+        thresholds.maxAvgRateRatio = 1.05;
+        thresholds.minAvgRateRatio = 1.00;
+    }
+}
+
+inline mfxF64 criticalRatio(mfxF64 maxRatio)
+{
+    // limit overrun more than twice is critical
+    // example: for maxRatio = 1.4 (40% limit), 80% excess is critical
+    return (maxRatio > 1.0) ? (2.0*maxRatio - 1.0) : maxRatio;
+}
+
 mfxStatus LookAheadBrc2::Init(MfxVideoParam  & video)
 {
     mfxExtCodingOptionDDI const & extDdi  = GetExtBufferRef(video);
     mfxExtCodingOption2   const & extOpt2 = GetExtBufferRef(video);
     mfxExtCodingOption3   const & extOpt3 = GetExtBufferRef(video);
 
+    sBrcTarget brcTarget = {0, MFX_BRC_ACCURACY_MEDIUM}; // will be added in API
```

Comment:
Remove comment about API... and add something to API?

Reply:
Done